### PR TITLE
Adding support for AI model routing with Not Diamond.

### PR DESCRIPTION
**Pull request by acompa**
Based off of the existing OpenAI integration. 

This PR is a WIP for the moment while I update the docs and run integration tests. But feel free to review if you'd like!

### CODE REVIEW

File: .github/workflows/quality-check-dagster-notdiamond.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+name: quality-check
```
Comment (cmpadden) [inline .github/workflows/quality-check-dagster-notdiamond.yml:1]: For the quality check, and build and release workflows, you should be able to copy:

- `quality-check-dagster-contrib-modal.yml`
- `release-dagster-contrib-modal.yml`

It looks like you copied the template itself, whereas the above invoke it like so:

```
jobs:
  build-and-release-dagster-contrib-modal:
    uses: ./.github/workflows/template-release.yml
    with:
      library_name: dagster-contrib-modal
      working_directory: ./libraries/dagster-contrib-modal
    secrets: inherit
```

File: libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/version.py [hunk 2]
```diff
@@ -0,0 +1 @@
+__version__ = "1!0+dev"
```
Comment (cmpadden) [inline libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/version.py:1]: We can remove this file.

File: libraries/dagster-contrib-notdiamond/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+[tool.uv]
+dev-dependencies = [
+    "ruff",
+    "pytest",
+    "pyright>=1.1.386",
+]
+
+[tool.setuptools]
+packages = ["dagster_notdiamond"]
```
Comment (cmpadden) [inline libraries/dagster-contrib-notdiamond/pyproject.toml:20]:
```suggestion
packages = ["dagster_contrib_notdiamond"]
```

File: libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/__init__.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from dagster._core.libraries import DagsterLibraryRegistry
+
+from dagster_notdiamond.resources import (
```
Comment (cmpadden) [inline libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/__init__.py:3]:
```suggestion
from dagster_contrib_notdiamond.resources import (
```

File: libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/__init__.py [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+    NotDiamondResource as NotDiamondResource,
+    with_usage_metadata as with_usage_metadata,
+)
+from dagster_notdiamond.version import __version__
```
Comment (cmpadden) [inline libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/__init__.py:7]:
```suggestion
from dagster_contrib_notdiamond.version import __version__
```

File: libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/resources.py [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+                AssetSelection,
+                AssetSpec,
+                Definitions,
+                EnvVar,
+                MaterializeResult,
+                asset,
+                define_asset_job,
+                multi_asset,
+            )
+            from dagster_notdiamond import NotDiamondResource, with_usage_metadata
```
Comment (cmpadden) [inline libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/resources.py:70]:
```suggestion
            from dagster_contrib_notdiamond import NotDiamondResource, with_usage_metadata
```

File: libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/resources.py [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+    By configuring this Not Diamond resource, you can interact with the Not Diamond API
+    and log its usage metadata in the asset metadata.
+
+    Examples:
+        .. code-block:: python
+
+            import os
+
+            from dagster import AssetExecutionContext, Definitions, EnvVar, asset, define_asset_job
+            from dagster_notdiamond import NotDiamondResource
```
Comment (cmpadden) [inline libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/resources.py:157]:
```suggestion
            from dagster_contrib_notdiamond import NotDiamondResource
```

File: libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/resources.py [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+                    AssetExecutionContext,
+                    Definitions,
+                    EnvVar,
+                    GraphDefinition,
+                    OpExecutionContext,
+                    asset,
+                    define_asset_job,
+                    op,
+                )
+                from dagster_notdiamond import NotDiamondResource
```
Comment (cmpadden) [inline libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/resources.py:244]:
```suggestion
                from dagster_contrib_notdiamond import NotDiamondResource
```

File: libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/resources.py [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+                    AssetKey,
+                    AssetSpec,
+                    Definitions,
+                    EnvVar,
+                    MaterializeResult,
+                    asset,
+                    define_asset_job,
+                    multi_asset,
+                )
+                from dagster_notdiamond import NotDiamondResource
```
Comment (cmpadden) [inline libraries/dagster-contrib-notdiamond/dagster_contrib_notdiamond/resources.py:314]:
```suggestion
                from dagster_contrib_notdiamond import NotDiamondResource
```

**Review by cmpadden — CHANGES_REQUESTED** [review]
Looking really good!

Minor change requests:

- Using the templates in GitHub actions
- Updating references to package name as `dagster_contrib_notdiamond`

Thanks again for taking the time to contribute, and work with the new integrations repo. Let me know if anything has been confusing.

File: libraries/dagster-contrib-notdiamond/README.md [hunk 10]
```diff
@@ -0,0 +1,4 @@
+# dagster-contrib-notdiamond
+
+The docs for `dagster-contrib-notdiamond` can be found
+[here](https://docs.dagster.io/_apidocs/libraries/dagster-notdiamond).
```
[the remark this answers is not in the document]
Comment (acompa) [reply libraries/dagster-contrib-notdiamond/README.md:4]: Thanks for letting me know! I don't have strong theories at the moment on these docs, but can draft something in the same spirit as Modal's contributed docs.

**Comment by acompa** [comment]
Okay @cmpadden - think I'm ready for a second set of eyes here. The method docstrings and README might have some inconsistencies, fyi!

**Comment by cmpadden** [comment]
> Okay @cmpadden - think I'm ready for a second set of eyes here. The method docstrings and README might have some inconsistencies, fyi!

This is looking really fantastic.

I'm attempting to run the unit tests locally, and am getting a weird setuptools error:

```
FAILED dagster_contrib_notdiamond_tests/test_resources.py::test_notdiamond_client_with_config - importlib.metadata.PackageNotFoundError: No package metadata was found for dagster-con...
```

Still looking into it, but if you have any ideas, please share!

**Comment by cmpadden** [comment]
Oh! It looks like the `dagster-openai` integration uses a `mock` library, and not the version that we're using here, that might be it!
```
from mock import ANY, MagicMock, patch
```

**Comment by acompa** [comment]
> Oh! It looks like the `dagster-openai` integration uses a `mock` library, and not the version that we're using here. Not sure why that's the case. Hmph.
> 
> ```
> from mock import ANY, MagicMock, patch
> ```

Yeah, and actually this library is not independently supported in modern Python versions. So I'm using`unittest.mock` here instead.

Issue seems to be caused by my forgetting of a mock call parameter. I've added it back, and all tests pass again locally.

**Review by cmpadden — APPROVED** [review]
Checks are running green--will follow up with you in Slack on next steps. 🤝

**Pull request merged**